source/ShipSelection.cpp

```diff
@@ -0,0 +1,231 @@
+/* ShipSelection.cpp
+Copyright (c) 2023 by Dave Flowers
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ShipSelection.h"
+
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Ship.h"
+
+#include <SDL2/SDL.h>
+
+using namespace std;
+
+namespace {
+	bool CanShowInSidebar(const Ship &ship, const Planet *here)
+	{
+		return ship.GetPlanet() == here;
+	}
+}
+
+
+
+ShipSelection::ShipSelection(PlayerInfo &player)
+	: player(player), selectedShip(player.Flagship())
+{
+	if(selectedShip)
+		allSelected.insert(selectedShip);
+}
+
+
+
+bool ShipSelection::Has(Ship *ship) const
+{
+	return allSelected.count(ship);
+}
+
+
+
+bool ShipSelection::HasMany() const
+{
+	return allSelected.size() > 1;
+}
+
+
+
+Ship *ShipSelection::Find(int count)
+{
+	if(!selectedShip)
+	{
+		Set(player.Flagship());
+		return nullptr;
+	}
+
+	// Find the currently selected ship in the list.
+	vector<shared_ptr<Ship>>::const_iterator it = player.Ships().begin();
+	for( ; it != player.Ships().end(); ++it)
+		if(it->get() == selectedShip)
+			break;
```

Comment:
```suggestion
	vector<shared_ptr<Ship>>::const_iterator it = player.Ships().find(selectedShip);
```
wouldn't this be the same? (dunno if the function is like that or `find(player.Ships.begin(), player.Ships().end(), slectedShip);`)

Reply:
Well, player.Ships() is a vector, which doesn't have a find method.  Also, selectedShip is a Ship *, not a shared_ptr<Ship>.  But I guess I could figure out how to use find with a lamda.

Reply:
a lambda would be nice yes,